[Test] Make iframe-zoom-nested.html WPT test less flaky

mdubet opened the pull request:
#### 4d172df73828e6c6f735795468f8dce5628dd6ba
<pre>
[Test] Make iframe-zoom-nested.html WPT test less flaky
<a href="https://bugs.webkit.org/show_bug.cgi?id=304258">https://bugs.webkit.org/show_bug.cgi?id=304258</a>

Reviewed by Tim Nguyen.

This test is using iframes (nested-iframe.html) of iframes (leaf.html) :
iframe loading via &quot;.src = &quot; is async,
we need to send a message to be sure to wait for the 4 leaf iframes to load
before capturing the rendering.

* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/iframe-zoom-nested.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/resources/leaf.html:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/platform/win/TestExpectations:

Canonical link: <a href="https://commits.webkit.org/304857@main">https://commits.webkit.org/304857@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bf121782db14b81872ba93b2f3d60f685c3c4259

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136643 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9002 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47930 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144367 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/89616 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7d066122-9fbb-4f88-b43d-a7c8eba8d1a3) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9703 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8848 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104485 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/75061 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f99aba7a-5e5f-4de2-a2ec-383ceb7d594b) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139588 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7085 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122434 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85322 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d0d47c2d-607f-463d-aac7-62dc1a00ee6d) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6729 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4412 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4963 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116044 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40623 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147126 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8686 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41197 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112838 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8704 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7306 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113168 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28754 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6652 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118727 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/62796 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8734 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36781 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8454 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8674 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8526 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/resources/leaf.html:
```diff
@@ -21,4 +21,10 @@
   document.body.style.setProperty("--scale", parseFloat(params.get("scale")));
 }
+addEventListener(
+    "load",
```
I'm not sure how this helps, we already waiting for the load event in all the frames. Seems like perhaps it just adds asynchronicity, basically equivalent to a zero-delay timer in original code.

In fact, this makes the test flaky in a different way, as it can now complete before the leaf_loaded messages are delivered. A test completes when the document and frames are done loading, and doesn't wait for messages.

mdubet replied:
Are you aware of `reftest-wait` ? If you do, I don't understand your comment. If not, please read about it and let me know if your comment stand ?

 https://web-platform-tests.org/writing-tests/reftests.html#controlling-when-comparison-occurs

aproskuryakov replied:
I'm not aware of `reftest-wait`, but also don't have any particular reason to believe that it affects what happens in `run-webkit-tests`. What I'm talking about is how WebKitTestRunner decides when to finish, and this test in particular doesn't need anything asynchronous, this is all just normal page load completion.

aproskuryakov replied:
What's definitely buggy in this test is that `load` even handlers are installed after the fact – loads can complete before the handlers are installed if the main resource is delivered in chunks, split between the iframes and the script. We see such issues all the time.

mdubet replied:
`leaf.html` is not a test by itself

aproskuryakov replied:
Yes, I'm looking at `iframe-zoom-nested.html`. There are no `load` event handlers in `leaf.html`.

aproskuryakov replied:
I checked into `reftest-wait` a bit more. Looks like we implement something around it in `Tools/TestRunnerShared/WPTFunctions.cpp`. Still, I don't see how it's relevant, given that that WebKitTestRunner waits for all subframes to load with or without it.

mdubet replied:
> I'm not aware of `reftest-wait`, but also don't have any particular reason to believe that it affects what happens in `run-webkit-tests`....

Well it does. I don't think requesting changes on a PR when you don't know about the main mechanism the PR is using is very valuable ? 

Also, we don't run test exclusively through WKTR (example: `WebDriver`)

aproskuryakov replied:
I feel the you are annoyed, but you are making the test more buggy. The PR needs to be revised.

aproskuryakov commented on LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/iframe-zoom-nested.html:
```diff
@@ -26,24 +26,24 @@
+            if (data != "leaf_loaded") return;
+            loadedCount += 1;
+            if (loadedCount == waitFor) finishTest();
+            //console.log("receive message" + loadedCount)
```
Please don't land commented out logs.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/iframe-zoom-nested.html:
```diff
@@ -26,24 +26,24 @@
+        document.documentElement.classList.remove("reftest-wait");
+    };
+    let loadedCount = 0;
+    const waitFor = 4;
```
I would preserve the old logic that counted the iframes, rather than a hardcoded number

nt1m replied:
also, let's move the variables inside the load event listener, since you're not using them outside.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/iframe-zoom-nested.html:
```diff
@@ -26,24 +26,24 @@
+    <iframe id="zoom-top" class="zoom" src="resources/nested-iframe.html" scrolling="no"></iframe>
+    <iframe id="zoom-top-child" class="zoom" src="resources/nested-iframe.html?zoom=2" scrolling="no"></iframe>
+    
 <script>
```
I would unindent because it looks inconsistent with `<script>`

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/iframe-zoom-nested.html:
```diff
@@ -26,24 +26,24 @@
-</script>
+    let finishTest = () => {
+        document.documentElement.classList.remove("reftest-wait");
+    };
```
Please use the helper from reftest-wait.js, rather than introducing a new one

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/iframe-zoom-nested.html:
```diff
@@ -26,24 +26,24 @@
+    let loadedCount = 0;
+    const waitFor = 4;
+    addEventListener("load", () => {
+        addEventListener("message", ({ data }) => {
```
Could we add a comment to explain what's being done here?

`// We need to wait for the nested iframe (leaf.html) to load, in addition to the top-level iframe (nested-iframe.html).`

nt1m reviewed (approved):
LGTM modulo style comments, but please add more context in the commit message:
- something explaining the nested iframe situation: how the old test just guaranteed that the top-level iframe was parsed, but not the leaf to be loaded. And how we're fixing that.

- what command you used to test that flakiness was gone.

mdubet committed:
[Test] Make iframe-zoom-nested.html WPT test less flaky
https://bugs.webkit.org/show_bug.cgi?id=304258

Reviewed by Tim Nguyen.

This test is using iframes (nested-iframe.html) of iframes (leaf.html) :
iframe loading via ".src = " is async,
we need to send a message to be sure to wait for the 4 leaf iframes to load
before capturing the rendering.

* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/iframe-zoom-nested.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/resources/leaf.html:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/platform/win/TestExpectations:

Canonical link: https://commits.webkit.org/304857@main
